resources/overlay_plugin_api.ts

```diff
@@ -0,0 +1,265 @@
+  callOverlayHandlerOverride?: IOverlayHandler;
+  dispatchOverlayEventOverride: typeof processEvent;
+}): void => {
+  const {
+    addOverlayListenerOverride,
+    removeOverlayListenerOverride,
+    callOverlayHandlerOverride,
+    dispatchOverlayEventOverride,
+  } = override;
+  if (addOverlayListenerOverride)
+    overrides.addOverlayListenerOverride = addOverlayListener;
+  if (removeOverlayListenerOverride)
+    overrides.removeOverlayListenerOverride = removeOverlayListener;
+  if (callOverlayHandlerOverride)
+    overrides.callOverlayHandlerOverride = callOverlayHandler;
+  if (dispatchOverlayEventOverride)
+    overrides.dispatchOverlayEvent = dispatchOverlayEventOverride;
+};
+
+if (typeof window !== 'undefined') {
+  wsUrl = /[\?&]OVERLAY_WS=([^&]+)/.exec(window.location.href);
+  if (wsUrl) {
+    const connectWs = function() {
+      ws = new WebSocket(wsUrl?.[1] as string);
+
+      ws.addEventListener('error', (e) => {
+        console.error(e);
+      });
+
+      ws.addEventListener('open', () => {
+        console.log('Connected!');
+
+        const q = queue;
+        queue = [];
+
+        sendMessage({
+          call: 'subscribe',
+          events: Object.keys(subscribers),
+        });
+
+        for (const msg of q) {
+          if (!Array.isArray(msg))
+            sendMessage(msg);
+        }
+      });
+
+      ws.addEventListener('message', (_msg) => {
+        try {
+          const msg = JSON.parse(_msg.data) as EventParameter & {rseq?: number};
+
+          if (msg.rseq !== undefined && responsePromises[msg.rseq]) {
+            responsePromises[msg.rseq]?.(msg);
+            delete responsePromises[msg.rseq];
+          } else {
+            processEvent(msg);
+          }
+        } catch (e) {
+          console.error('Invalid message received: ', _msg);
+          return;
+        }
+      });
+
+      ws.addEventListener('close', () => {
+        queue = [];
+
+        console.log('Trying to reconnect...');
+        // Don't spam the server with retries.
+        setTimeout(() => {
+          connectWs();
+        }, 300);
+      });
+    };
+
+    connectWs();
+  } else {
+    const waitForApi = function() {
+      if (!window.OverlayPluginApi || !window.OverlayPluginApi.ready) {
+        setTimeout(waitForApi, 300);
+        return;
+      }
+
+      const q = queue;
+      queue = [];
+
+      window.__OverlayCallback = processEvent;
+
+      sendMessage({
+        call: 'subscribe',
+        events: Object.keys(subscribers),
+      });
+
+      for (const item of q) {
+        if (Array.isArray(item))
+          sendMessage(item[0], item[1]);
+      }
+    };
+
+    waitForApi();
+  }
+
+  // Assign in `window` object to ensure backward compatiability
+  window.addOverlayListener = addOverlayListener;
+  window.removeOverlayListener = removeOverlayListener;
+  window.callOverlayHandler = callOverlayHandler;
+  window.dispatchOverlayEvent = processEvent;
```

Comment:
Is it possible to only do this for user trigger files?

Reply:
Ok, I'll move them in user_config.js

Reply:
But the `dispatchOverlayEvent` should be still in this file, right ?

Reply:
Sorry for belated reply, as I'm a little bit busy/exhausted these days.  I am not actually sure what `dispatchOverlayEvent` is.  It doesn't look like it is called in cactbot, from what I can tell.  I also can't see that it's called via OverlayPlugin either.  So, maybe it can also be handled in the same way (i.e. exported, and only on window)?

Also, `window` is a global, so even if this is only set inside of user_config, that will still be visible to other files.  Is it possible to set and un-set it off of window around the `eval` blocks?

Reply:
> Sorry for belated reply, as I'm a little bit busy/exhausted these days. I am not actually sure what `dispatchOverlayEvent` is. It doesn't look like it is called in cactbot, from what I can tell. I also can't see that it's called via OverlayPlugin either. So, maybe it can also be handled in the same way (i.e. exported, and only on window)?
> 
> Also, `window` is a global, so even if this is only set inside of user_config, that will still be visible to other files. Is it possible to set and un-set it off of window around the `eval` blocks?

Oh, I thought it would be called by OverlayPlugin or something in Cactbot's C# backend, 'cause it looks like an event dispatcher callback.
Okay, I'll change like what you said.